py/desiutil/bitmask.py

```diff
@@ -0,0 +1,162 @@
+"""
+desiutil.maskbits
+=================
+
+Mask bits for the spectro pipeline.
+
+Individual packages will define their own mask bits and use this as a utility
+access wrapper.  Typical users will not need to construct BitMask objects on
+their own.
+
+Stephen Bailey, Lawrence Berkeley National Lab
+Fall 2015
+
+Example::
+
+    #- Creating a BitMask
+    from desiutil.bitmask import BitMask
+
+    import yaml
+    _bitdefs = yaml.load('''
+    ccdmask:
+        - [BAD,       0, "Pre-determined bad pixel (any reason)"]
+        - [HOT,       1, "Hot pixel"]
+        - [DEAD,      2, "Dead pixel"]
+        - [SATURATED, 3, "Saturated pixel from object"]
+        - [COSMIC,    4, "Cosmic ray"]
+    ''')
+    ccdmask = BitMask('ccdmask', _bitdefs)
+
+    #- Accessing the mask
+    ccdmask.COSMIC | ccdmask.SATURATED  #- 2**4 + 2**3
+    ccdmask.mask('COSMIC')     #- 2**4, same as ccdmask.COSMIC
+    ccdmask.mask(4)            #- 2**4, same as ccdmask.COSMIC
+    ccdmask.COSMIC             #- 2**4, same as ccdmask.mask('COSMIC')
+    ccdmask.bitnum('COSMIC')   #- 4
+    ccdmask.bitname(4)         #- 'COSMIC'
+    ccdmask.names()            #- ['BAD', 'HOT', 'DEAD', 'SATURATED', 'COSMIC']
+    ccdmask.names(3)           #- ['BAD', 'HOT']
+    ccdmask.comment(0)         #- "Pre-determined bad pixel (any reason)"
+    ccdmask.comment('COSMIC')  #- "Cosmic ray"
+"""
+
+
+class _MaskBit(int):
+    """
+    A single mask bit.  Subclasses int to act like an int, but allows the
+    ability to extend with blat.name, blat.comment, blat.mask, blat.bitnum.
+    """
+    def __new__(cls, name, bitnum, comment, extra=dict()):
+        self = super(_MaskBit, cls).__new__(cls, 2**bitnum)
+        self.name = name
+        self.bitnum = bitnum
+        self.mask = 2**bitnum
+        self.comment = comment
+        for key, value in extra.items():
+            assert key not in (
+                'bitlength', 'conjugate', 'real', 'imag',
+                'numerator', 'denominator'), \
+                "key '{}' already in use by int objects".format(key)
+            self.__dict__[key] = value
+        return self
+
+    def __str__(self):
+        return '{:16s} bit {} mask 0x{:X} - {}'.format(
+            self.name, self.bitnum, self.mask, self.comment)
+
+
+#- Class to provide mask bit utility functions
+class BitMask(object):
+    """BitMask object.
+    """
```

Comment:
It could be useful to have `help(ccdmask)` list the individual bits and their comment strings, rather than a generic `BitMask object` docstring.  One way to accomplish this is to replace `class BitMask` with a function `def createBitMask(name, bitdefs)` that returns
```
type(name, (), class_dict)
```
for an appropriately constructed per-bitmask class dictionary, including a `__doc__` attribute.